lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +182,212 @@ class Loader {
+      'string, SharedArrayBuffer, or Uint8Array',
+      'loader load',
+      'source',
+      source
+    );
+
+    return {
+      format,
+      source,
+    };
   }
 
-  runGlobalPreloadCode() {
-    if (!this._getGlobalPreloadCode) {
-      return;
-    }
-    const preloadCode = this._getGlobalPreloadCode();
-    if (preloadCode === null) {
-      return;
-    }
+  preload() {
+    const count = this.#globalPreloaders.length;
+    if (!count) return;
 
-    if (typeof preloadCode !== 'string') {
-      throw new ERR_INVALID_RETURN_VALUE(
-        'string', 'loader getGlobalPreloadCode', preloadCode);
-    }
-    const { compileFunction } = require('vm');
-    const preloadInit = compileFunction(preloadCode, ['getBuiltin'], {
-      filename: '<preload>',
-    });
-    const { NativeModule } = require('internal/bootstrap/loaders');
-
-    preloadInit.call(globalThis, (builtinName) => {
-      if (NativeModule.canBeRequiredByUsers(builtinName)) {
-        return require(builtinName);
+    for (let i = count - 1; i > -1; i--) {
+      const preloadCode = this.#globalPreloaders[i]();
+      if (preloadCode === null) return;
+
+      if (typeof preloadCode !== 'string') {
+        throw new ERR_INVALID_RETURN_VALUE(
+          'string',
+          'loader globalPreloadCode',
+          preloadCode,
+        );
       }
-      throw new ERR_INVALID_ARG_VALUE('builtinName', builtinName);
-    });
+      const { compileFunction } = require('vm');
+      const preloadInit = compileFunction(
+        preloadCode,
+        ['getBuiltin'],
+        {
+          filename: '<preload>',
+        }
+      );
+      const { NativeModule } = require('internal/bootstrap/loaders');
+
+      preloadInit.call(globalThis, (builtinName) => {
+        if (NativeModule.canBeRequiredByUsers(builtinName)) {
+          return require(builtinName);
+        }
+        throw new ERR_INVALID_ARG_VALUE('builtinName', builtinName);
+      });
+    }
   }
 
-  async getModuleJob(specifier, parentURL) {
-    const url = await this.resolve(specifier, parentURL);
-    const format = await this.getFormat(url);
-    let job = this.moduleMap.get(url);
-    // CommonJS will set functions for lazy job evaluation.
-    if (typeof job === 'function')
-      this.moduleMap.set(url, job = job());
-    if (job !== undefined)
-      return job;
+  /**
+   * Resolve the location of the module.
+   *
+   * The internals of this WILL change when chaining is implemented,
+   * depending on the resolution/consensus from #36954
+   * @param {*} originalSpecifier - The specified URL path of the module to be
+   * resolved
+   * @param {String} parentURL - The URL path of the module's parent
+   * @returns {{ url: String }}
+   */
+  async resolve(originalSpecifier, parentURL) {
+    const conditions = DEFAULT_CONDITIONS;
 
-    if (!translators.has(format))
-      throw new ERR_UNKNOWN_MODULE_FORMAT(format);
+    const [ defaultResolver, customResolver ] = this.#resolvers;
 
```

Comment:
I did. `resolvers` is an array and needs to stay an array (or array-like); the lint error seems nonsense. Is it suggesting something like

```
const {
  0: defaultResolver,
  1: customResolver,
} = this.#resolvers;
```

That _technically_ works, but…why?

Reply:
That's to avoid unsafe array iteration. When using array destructuring assignment, it calls `this.#resolver[Symbol.iterator]` which could be mutated in userland. We try to make internal code not rely on that, to Node.js throwing or going into infinite loops if users temper with built-in objects.

```js
const array = [1 ,2, 3];
delete Array.prototype[Symbol.iterator];

const [one, two] = array; // throws
const { 0: one, 1: two } = array; // doesn't throw
```

Reply:
A user could not possibly do this by accident, so (not to be overly critical) it sounds like a case of a user shooting themselves in the foot and deserving the consequences…

Perhaps the end-user is the victim of a degenerate library, but they're likely gonna find out one way or another. Better to have a very in-your-face and prolific fireworks show than subtle wtf failures, no?

Not a discussion for a singular PR though—I'll update it to the indexed destructuring.